sound/soc/sof/topology.c

```diff
@@ -2019,6 +2042,108 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
 	return 0;
 }
 
+/*
+ * Sof mux/demux component
+ */
+static int sof_process_load_mux(struct snd_soc_component *scomp,
+				struct sof_ipc_comp_process *process,
+				struct snd_soc_tplg_private *private,
+				int *offset)
+{
+	struct sof_ipc_stream_map *sm =
+		(struct sof_ipc_stream_map *)((u8 *)process->data + (*offset));
+	u8 *pos = (u8 *)process->data + (*offset) +
+		sizeof(struct sof_ipc_stream_map);
+	struct sof_ipc_channel_map *cm_temp = NULL;
+	struct sof_ipc_channel_map *cm_temp_p;
+	u32 *cm_coeffs_temp = NULL;
+	u32 *cm_coeffs_temp_p;
+	int num_coeffs = 0;
+	size_t size;
+	int ret;
+	int i;
+
+	/* parse stream map */
+	ret = sof_parse_tokens(scomp, sm, stream_map_tokens,
+			       ARRAY_SIZE(stream_map_tokens), private->array,
+			       le32_to_cpu(private->size));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: mux stream map tokens failed %d\n",
+			private->size);
+		return ret;
+	}
+
+	/* reserve temp memory for variable amount of channel maps */
+	size = sizeof(struct sof_ipc_channel_map) * sm->num_ch_map;
+	cm_temp = kzalloc(size, GFP_KERNEL);
+	if (!cm_temp)
+		return -ENOMEM;
+
+	/* parse channel maps to temp memory cm_temp */
+	ret = sof_parse_token_sets(scomp, cm_temp, channel_map_tokens,
+				   ARRAY_SIZE(channel_map_tokens),
+				   private->array,
+				   le32_to_cpu(private->size),
+				   sm->num_ch_map,
+				   sizeof(struct sof_ipc_channel_map));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: mux channel map tokens %d\n",
+			private->size);
+		goto out;
+	}
+
+	/* calculate total amount of coeffs i.e. all bits set in all masks */
+	cm_temp_p = cm_temp;
+	for (i = 0; i < sm->num_ch_map; i++, cm_temp_p++)
+		num_coeffs += __sw_hweight32(cm_temp_p->ch_mask);
+
+	/* reserve temp memory for variable amount of coeffs */
+	size = sizeof(u32) * num_coeffs;
+	cm_coeffs_temp = kzalloc(size, GFP_KERNEL);
+	if (!cm_coeffs_temp) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	/* parse coeffs to temp memory cm_coeffs */
+	ret = sof_parse_token_sets(scomp, cm_coeffs_temp,
+				   channel_map_coeff_tokens,
+				   ARRAY_SIZE(channel_map_coeff_tokens),
+				   private->array,
+				   le32_to_cpu(private->size),
+				   num_coeffs,
+				   sizeof(u32));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: mux stream tokens %d\n",
+			private->size);
+		goto out;
+	}
+
+	/* update offset as num_coeffs is still valid for total amount */
+	*offset += sizeof(struct sof_ipc_stream_map) +
+		sm->num_ch_map * sizeof(struct sof_ipc_channel_map) +
+		num_coeffs * sizeof(u32);
+
+	/* copy map and coeff data from temp areas to process struct */
+	num_coeffs = 0;
+	cm_temp_p = cm_temp;
+	cm_coeffs_temp_p = cm_coeffs_temp;
+	for (i = 0; i < sm->num_ch_map; i++) {
+		memcpy(pos, cm_temp_p, sizeof(struct sof_ipc_channel_map));
```

Comment:
I think we can remove cm_temp_p and just use cm_temp + i * sizeof(*cm_temp) here?

Reply:
hmm ok. it might get little bit messier, but yeah I'll try it.

Reply:
that would mean a multiplication on each loop iteration :-) Yes, that's micro-optimisation and we have huge beefy CPUs, but how about something like
```
for (i = 0, cm_temp_p = cm_temp;
      i < sm->num_ch_map;
      i++, cm_temp_p++)
```

Reply:
yes, it is just my personal preference to not include multiple "end statement" to for loops, I personally many times forget to see that and wonder who is updating some variable. And maybe even more stylistic preference for myself...